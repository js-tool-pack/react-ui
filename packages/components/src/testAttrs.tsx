import { fireEvent, render } from '@testing-library/react';
import type { PropsBase } from '@pkg/shared';
import React from 'react';

export function testAttrs<
  T extends { attrs?: Partial<React.HTMLAttributes<HTMLElement>> },
>(component: React.FC<T>, displayName?: string) {
  const name = displayName || component.displayName || component.name;
  test(name ? `${name} attrs` : `attrs`, () => {
    const onClick = jest.fn();

    const FC = component as React.FC<PropsBase>;
    const { container } = render(
      <FC
        attrs={{
          style: { width: '20px' },
          className: 'foo',
          role: 'note',
          onClick,
        }}
      />,
    );

    expect(container.firstChild).toHaveClass('foo');
    expect(container.firstChild).toHaveAttribute('role', 'note');
    expect(container.firstChild).toHaveStyle({ width: '20px' });

    expect(onClick).not.toBeCalled();
    fireEvent.click(container.firstChild!);
    expect(onClick).toBeCalled();
  });
}
