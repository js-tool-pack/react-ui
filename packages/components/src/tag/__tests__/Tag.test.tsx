import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { Tag } from '..';
import { testAttrs } from '~/testAttrs';
import Demo from '../demo/basic';

describe('Tag', () => {
  testAttrs(Tag);

  test('demo', () => {
    const { container } = render(<Demo />);
    expect(container.firstChild).toMatchSnapshot();
  });

  test('bordered', () => {
    expect(render(<Tag />).container.firstChild).toHaveClass('t-tag--bordered');
    expect(
      render(<Tag bordered={false} />).container.firstChild,
    ).not.toHaveClass('t-tag--bordered');
  });

  test('round', () => {
    expect(render(<Tag round />).container.firstChild).toHaveClass(
      't-tag--round',
    );
    expect(render(<Tag />).container.firstChild).not.toHaveClass(
      't-tag--round',
    );
  });

  test('size', () => {
    expect(render(<Tag size="small" />).container.firstChild).toHaveClass(
      't--size-sm',
    );
    expect(render(<Tag size="medium" />).container.firstChild).toHaveClass(
      't--size-m',
    );
    expect(render(<Tag />).container.firstChild).toHaveClass('t--size-m');
    expect(render(<Tag size="large" />).container.firstChild).toHaveClass(
      't--size-lg',
    );
  });

  test('icon', () => {
    expect(
      render(<Tag icon={'tag'} size="small" />).container.firstChild,
    ).toMatchSnapshot();
  });

  describe('checkable', () => {
    test('unable', () => {
      const onChange = jest.fn();
      const { container } = render(
        <Tag onChange={onChange} checkable={false} />,
      );
      expect(container.firstChild).not.toHaveClass('t-tag--checkable');
      expect(container.firstChild).not.toHaveClass('t-tag--checked');
      expect(onChange).not.toBeCalled();

      fireEvent.click(container.firstChild!);
      expect(onChange).not.toBeCalled();
      expect(container.firstChild).not.toHaveClass('t-tag--checked');
    });

    test('enable', () => {
      const onChange = jest.fn();
      const { container } = render(<Tag onChange={onChange} checkable />);

      expect(container.firstChild).toHaveClass('t-tag--checkable');
      expect(container.firstChild).not.toHaveClass('t-tag--checked');
      expect(onChange).not.toBeCalled();

      fireEvent.click(container.firstChild!);
      expect(onChange).toBeCalled();
      expect(onChange.mock.calls).toEqual([[true]]);
      expect(container.firstChild).toHaveClass('t-tag--checked');

      fireEvent.click(container.firstChild!);
      expect(onChange.mock.calls).toEqual([[true], [false]]);
      expect(container.firstChild).not.toHaveClass('t-tag--checked');
    });

    test('closeBtnAttrs', () => {
      const onChange = jest.fn();
      const { container } = render(
        <Tag onChange={onChange} closeable closeBtnAttrs={{ tabIndex: -1 }} />,
      );
      expect(container.querySelector('button')).toHaveAttribute(
        'tabindex',
        '-1',
      );
    });

    test('checked', () => {
      const onChange = jest.fn();
      const { container } = render(
        <Tag onChange={onChange} checked checkable />,
      );

      expect(container.firstChild).toHaveClass(
        't-tag--checkable t-tag--checked',
      );
      expect(onChange).not.toBeCalled();

      fireEvent.click(container.firstChild!);
      expect(onChange).toBeCalled();
      expect(onChange.mock.calls).toEqual([[false]]);
      expect(container.firstChild).not.toHaveClass('t-tag--checked');

      fireEvent.click(container.firstChild!);
      expect(onChange.mock.calls).toEqual([[false], [true]]);
      expect(container.firstChild).toHaveClass('t-tag--checked');
    });

    test('checked uncheckable', () => {
      const onChange = jest.fn();
      const { container } = render(<Tag onChange={onChange} checked />);

      expect(container.firstChild).toHaveClass('t-tag--checked');
      expect(container.firstChild).not.toHaveClass('t-tag--checkable');
      expect(onChange).not.toBeCalled();

      fireEvent.click(container.firstChild!);
      expect(onChange).not.toBeCalled();
      expect(container.firstChild).toHaveClass('t-tag--checked');
    });
  });

  describe('closeable', () => {
    test('basic', () => {
      const onClose = jest.fn();
      const { container } = render(
        <Tag closeable onClose={onClose}>
          tag
        </Tag>,
      );

      expect(onClose).not.toBeCalled();
      expect(container.firstChild).toMatchSnapshot();
      expect(container.firstChild).not.toBeNull();

      fireEvent.click(container.querySelector('.t-tag__close')!);
      expect(onClose).toBeCalled();
      expect(container.firstChild).toMatchSnapshot();
      expect(onClose.mock.calls[0][0]._reactName).toBe('onClick');
      expect(container.firstChild).toBeNull();
    });

    test('preventDefault', () => {
      const onClose = jest.fn((e: React.MouseEvent) => e.preventDefault());
      const { container } = render(
        <Tag closeable onClose={onClose}>
          tag
        </Tag>,
      );

      expect(onClose).not.toBeCalled();
      expect(container.firstChild).not.toBeNull();

      fireEvent.click(container.querySelector('.t-tag__close')!);
      expect(onClose).toBeCalled();
      expect(container.firstChild).not.toBeNull();
    });
  });

  describe('disabled', () => {
    test('basic', () => {
      expect(render(<Tag />).container.firstChild).not.toHaveClass(
        't-tag--disabled',
      );
      expect(render(<Tag disabled />).container.firstChild).toHaveClass(
        't-tag--disabled',
      );
    });

    test('checkable', () => {
      const onChange = jest.fn();
      const { container } = render(
        <Tag onChange={onChange} disabled checkable />,
      );

      expect(container.firstChild).toHaveClass('t-tag--checkable');
      expect(container.firstChild).not.toHaveClass('t-tag--checked');
      expect(onChange).not.toBeCalled();

      fireEvent.click(container.firstChild!);
      expect(onChange).not.toBeCalled();
      expect(container.firstChild).not.toHaveClass('t-tag--checked');
    });

    test('closeable', () => {
      const onClose = jest.fn();
      const { container } = render(
        <Tag closeable disabled onClose={onClose}>
          tag
        </Tag>,
      );

      expect(onClose).not.toBeCalled();
      expect(container.firstChild).not.toBeNull();

      fireEvent.click(container.querySelector('.t-tag__close')!);
      expect(onClose).not.toBeCalled();
      expect(container.firstChild).not.toBeNull();
    });

    test('关闭按钮拦截外部点击事件', () => {
      const onWindowClick = jest.fn();
      const onClick = jest.fn();
      const { container } = render(
        // eslint-disable-next-line jsx-a11y/click-events-have-key-events,jsx-a11y/no-static-element-interactions
        <div onClick={onClick}>
          <Tag closeable>tag</Tag>
        </div>,
      );

      window.addEventListener('click', onWindowClick);

      expect(onClick).not.toBeCalled();
      expect(onWindowClick).not.toBeCalled();
      expect(container.firstChild!.firstChild).not.toBeNull();

      fireEvent.click(container.querySelector('.t-tag__close')!);
      expect(onClick).not.toBeCalled();
      expect(onWindowClick).not.toBeCalled();
      expect(container.firstChild!.firstChild).toBeNull();
    });
  });
});
