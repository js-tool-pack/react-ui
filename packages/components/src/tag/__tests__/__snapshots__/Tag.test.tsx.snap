// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tag closeable basic 1`] = `
<div
  class="t-tag t--size-m t-tag--primary t-tag--bordered"
>
  <span
    class="t-tag__content"
  >
    tag
  </span>
  <button
    class="t-button t-tag__close t-button--type-primary t--size-sm t-button--plain-text t-button--icon-l"
  >
    <span>
      <i
        class="t-icon"
        role="img"
      >
        <svg
          viewBox="0 0 1024 1024"
        >
          <path
            d="M574.56865938 513.72328935l366.30945353-364.40941657c17.3212673-17.23289347 17.4096411-45.15901828 0.17674762-62.48028556-17.27708037-17.36545419-45.15901828-17.4096411-62.48028556-0.17674764l-366.39782734 364.45360348-363.21637009-364.32104275C131.77167097 69.55650684 103.80135925 69.46813303 86.48009197 86.70102651 69.20301159 103.93391997 69.15882467 131.90423169 86.39171815 149.22549897l363.12799629 364.23266894L83.38700853 877.64664993c-17.3212673 17.23289347-17.4096411 45.15901828-0.17674763 62.48028558 8.61644674 8.66063363 19.97248167 13.03513737 31.32851659 13.03513736 11.26766112 0 22.53532223-4.28612991 31.15176897-12.85838974l366.22107973-364.27685584 366.53038806 367.63506071c8.61644674 8.66063363 19.92829476 12.99095045 31.2843297 12.99095047 11.31184801 0 22.57950913-4.33031682 31.19595588-12.90257666 17.27708037-17.23289347 17.3212673-45.15901828 0.0883738-62.48028554L574.56865938 513.72328935z"
            fill="currentColor"
          />
        </svg>
      </i>
    </span>
  </button>
</div>
`;

exports[`Tag closeable basic 2`] = `null`;

exports[`Tag demo 1`] = `
<section
  class="t-space"
  style="--t-space-fill-ratio: 100%;"
>
  <div
    class="t-tag t--size-m t-tag--primary t-tag--bordered"
  >
    <span
      class="t-tag__content"
    >
      DEFAULT
    </span>
  </div>
  <div
    class="t-tag t--size-m t-tag--primary t-tag--bordered"
  >
    <span
      class="t-tag__content"
    >
      PRIMARY
    </span>
  </div>
  <div
    class="t-tag t--size-m t-tag--success t-tag--bordered"
  >
    <span
      class="t-tag__content"
    >
      SUCCESS
    </span>
  </div>
  <div
    class="t-tag t--size-m t-tag--info t-tag--bordered"
  >
    <span
      class="t-tag__content"
    >
      INFO
    </span>
  </div>
  <div
    class="t-tag t--size-m t-tag--warning t-tag--bordered"
  >
    <span
      class="t-tag__content"
    >
      WARNING
    </span>
  </div>
  <div
    class="t-tag t--size-m t-tag--error t-tag--bordered"
  >
    <span
      class="t-tag__content"
    >
      ERROR
    </span>
  </div>
</section>
`;

exports[`Tag icon 1`] = `
<div
  class="t-tag t--size-sm t-tag--primary t-tag--bordered"
>
  <i
    class="t-icon t-tag__icon"
    role="img"
  >
    tag
  </i>
  <span
    class="t-tag__content"
  />
</div>
`;
